File: II-159-3/ThePriceIsRight.cs
using System;
using System.Collections;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

public class ThePriceIsRight 
{
	public int[] howManyReveals( int[] prices ) 
	{
		int n = prices.Length;
		int[] A = new int[n];
		int[] B = new int[n];
		int max_i = 0;
		int max_count = 0;
		for( int i = 0; i < n; i++ )
		{
			A[i] = B[i] = 1;
			for( int j = 0; j < i; j++ )
				if ( prices[i] > prices[j] )
					if ( A[i] == A[j] + 1 )
						B[i] += B[j];
					else if ( A[i] < A[j] + 1 )
					{
						A[i] = A[j] + 1;
						B[i] = B[j];
						if ( A[max_i] < A[i] )
							max_i = i;
					}
		}
		for( int i = 0; i < n; i++ )
			if ( A[i] == A[max_i] )
				max_count += B[i];
		return new int[] {A[max_i], max_count};
	}

	NumberFormatInfo nfi = new CultureInfo( "en-US", false ).NumberFormat;

// BEGIN CUT HERE
	public static void Main(string[] args)
	{
		DateTime t1 = DateTime.Now;
		eq(1,(new ThePriceIsRight()).howManyReveals(new int[] {30,10,20,40,50}),new int[] { 4,  1 });
		eq(2,(new ThePriceIsRight()).howManyReveals(new int[] {39,88,67,5,69,87,82,64,58,61}),new int[] { 4,  2 });
		eq(3,(new ThePriceIsRight()).howManyReveals(new int[] {1,2,3,4,5,6,7,8,9,10}),new int[] { 10,  1 });
		eq(4,(new ThePriceIsRight()).howManyReveals(new int[] {10,9,8,7,6,5,4,3,2,1}),new int[] { 1,  10 });
		eq(5,(new ThePriceIsRight()).howManyReveals(new int[] {29,31,73,70,14,5,6,34,53,30,15,86}),new int[] { 5,  2 });
		eq(6,(new ThePriceIsRight()).howManyReveals(new int[] {100,99,1,2,3}),new int[] { 3,  1 });
		eq(7,(new ThePriceIsRight()).howManyReveals(new int[] {10,20,11,12}),new int[] { 3,  1 });
		DateTime t2 = DateTime.Now;
		TimeSpan t = t2 - t1;
		Console.WriteLine( "{0} sec", t.TotalMilliseconds / 1000 ); 
	}
	private static void eq(int n, int a, int b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, double a, double b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, char a, char b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
	}
	private static void eq(int n, long a, long b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, bool a, bool b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, string a, string b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, int[] a, int[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, double[] a, double[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, long[] a, long[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, string[] a, string[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void print(int a)
	{
		Console.Write(a+" ");
	}
	private static void print(string s)
	{
		Console.Write("\""+s+"\" ");
	}
	private static void print(int[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(double[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(long[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(string[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write("\""+rs[i]+"\"");
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void nl()
	{
		Console.WriteLine();
	}
// END CUT HERE
}

File: II-159-3/ThePriceIsRight.cpp
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE
typedef vector<int> vi;
#define fora(i,v) for(int i = 0; i < (v).size(); i++)
#define all(v) (v).begin(), (v).end()
struct ThePriceIsRight
{
	vi howManyReveals( vi prices )
	{
		vi result; int n = prices.size();
		vi A(n, 1), B(n, 1);
		for( int i = 0; i < n; ++i )
			for( int j = 0; j < i; ++j )
				if ( prices[i] > prices[j] )
					if ( A[i] <= A[j] )
					{
						A[i] = A[j] + 1; B[i] = B[j];
					} else if ( A[i] == A[j] + 1 )
						B[i] += B[j];
		int k = *max_element( all(A) );
		result.push_back( k ); result.push_back( 0 );
		fora(i, A)
			if ( A[i] == k )
				result.back() += B[i];
		return result;
	}
	// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const long long& expected, const long long& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<vi> A0; vector<vi> A1;
	{ /*Test 0*/int _prices[] = 
	{30,10,20,40,50}
	; vi prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); int _result[] = 
	{ 4,  1 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 1*/int _prices[] = 
	{39,88,67,5,69,87,82,64,58,61}
	; vi prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); int _result[] = 
	{ 4,  2 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 2*/int _prices[] = 
	{1,2,3,4,5,6,7,8,9,10}
	; vi prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); int _result[] = 
	{ 10,  1 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 3*/int _prices[] = 
	{10,9,8,7,6,5,4,3,2,1}
	; vi prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); int _result[] = 
	{ 1,  10 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 4*/int _prices[] = 
	{29,31,73,70,14,5,6,34,53,30,15,86}
	; vi prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); int _result[] = 
	{ 5,  2 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 5*/int _prices[] = 
	{100,99,1,2,3}
	; vi prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); int _result[] = 
	{ 3,  1 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 6*/int _prices[] = 
	{10,20,11,12}
	; vi prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); int _result[] = 
	{ 3,  1 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A1[i], howManyReveals( A0[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	ThePriceIsRight ___test; 
	___test.run_test(); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE
